src/storage/swap.cc

```diff
@@ -16,35 +18,59 @@ std::shared_ptr<Swap> Swap::_GetSharedRef() {
 
 Swap::Swap(){
   std::cout << "Initialize Swap" <<std::endl;
-  mhistory_ = MemHistory::_GetSharedRef();
+  memory_history_ = MemHistory::_GetSharedRef();
+  memory_manager_ = MemoryManager::_GetSharedRef();
+  swap_lock_ = PTHREAD_RWLOCK_INITIALIZER;
+  for (int i = 0; i < NUMBER_OF_GPU; ++i){
+    locks_[i] = PTHREAD_RWLOCK_INITIALIZER;
+    free_memory_.push_back(0);
+  }
 }
 
 Swap::~Swap(){
   std::cout << "Destroy Swap" <<std::endl;
 }
 
 void Swap::SwapOut(unsigned required_memory, int device){
-  // Not implemented  
+  UpdateFree(device);
+  if (free_memory_[device] > required_memory) {
```

Comment:
This may not an accurate one. Instead, memory_manager_ should provide a function "TryAllocate(size_t size)". It will return true if and only if the function can allocate a continuous "size" memory. You can add a FIXME if you can't fix it in this pull request.